src/core/login-core.js

```diff
@@ -0,0 +1,219 @@
+                  token: tokenForUser(id.id)
+                };
+              });
+          }
+        });
+    });
+}
+
+export function addModerator(origemail, email, password) {
+  return knex("role")
+    .select("id")
+    .where("email", email)
+    .orWhere("email", email)
+    .then(rows => {
+      if (_.isEmpty(rows)) {
+        return knex("role")
+          .insert({ email: email, password: password })
+          .returning("id")
+          .then(result => {
+            const [id] = result;
+            const token = tokenForUser(id);
+            return sendTokenWithEmail(origemail, token);
+          });
+      } else {
+        return throwStatus(422, "User with the given email exists");
+      }
+    });
+}
+
+export function deleteModerator(id) {
+  return knex("role")
+    .where("id", id)
+    .del()
+    .then(result => {
+      if (result == 1) {
+        return throwStatus(200, "User deleted");
+      } else {
+        return throwStatus(404, "Moderator not found");
+      }
+    });
+}
+
+export function changePassword(password, oldpassword, user) {
+  return knex("role")
+    .select("password")
+    .where("id", user)
+    .then(pw => {
+      const [pass] = pw;
+      if (pass.password != oldpassword) {
+        return throwStatus(401, "Old password is not correct");
+      } else {
+        return knex("role")
+          .where("id", user)
+          .update({ password: password })
+          .returning("id")
+          .then(row => {
+            if (_.isEmpty(row)) {
+              return throwStatus(404, "There was an error");
+            } else {
+              return throwStatus(200, "Password changed");
+            }
+          });
+      }
+    });
+}
+
+export function activateAccount(password, user) {
+  return knex("role")
+    .where("id", user)
+    .update({ password: password, activated: true })
+    .returning("id")
+    .then(row => {
+      if (_.isEmpty(row)) {
+        return throwStatus(404, "There was an error");
+      } else {
+        return throwStatus(200, "Password changed and account activated");
+      }
+    });
+}
+
+export function promote(id) {
+  return knex("role")
+    .where("id", id)
+    .update({ admin: true })
+    .returning("id")
+    .then(row => {
+      if (_.isEmpty(row)) {
+        return throwStatus(404, "Moderator not found");
+      } else {
+        return throwStatus(200, "User promoted to admin");
+      }
+    });
+}
+
+export function demote(id) {
+  return knex("role")
+    .where("id", id)
+    .update({ admin: false })
+    .returning("id")
+    .then(row => {
+      if (_.isEmpty(row)) {
+        return throwStatus(404, "Moderator not found");
+      } else {
+        return throwStatus(200, "User demoted to moderator");
+      }
+    });
+}
+
+export function modList() {
+  return knex("role")
+    .select("id", "email", "activated", "admin")
+    .then(rows => {
+      if (_.isEmpty(rows)) {
+        return throwStatus(404, "No moderator list");
```

Comment:
Is "no moderators" actually an error state?

Reply:
Not necessarily. Will look into.